tesla_ver/data_uploading/data_uploading.py

```diff
@@ -0,0 +1,115 @@
+            Output("data-table-visualization", "data"),
+            Output("data-table-visualization", "columns"),
+        ],
+        [Input("upload", "contents"), Input("upload", "filename")],
+    )
+    def upload_data_parsing(contents, filename):
+        def upload_string_to_df(content):
+            """Parse a dash upload string into a single dataframe
+
+                Args:
+                    content (b64 string): string passed from the upload component
+            """
+            _, content_string = content.split(",")
+            decoded = b64decode(content_string)
+            fileish = StringIO(decoded.decode("utf-8"))
+            return pd.read_csv(fileish, sep="\t")
+
+        def parse_multiple_contents(contents, filenames):
+            """parses and merges multiple dataframes into a single selection
+
+            Args:
+                contents (list(b64 encoded string)): list of base64 encoded strings representing dataframes from dash upload
+                filenames (list(str)): list of file names corresponding to each dataframe
+            """
+            pass
+
+        df = pd.DataFrame()
+        if None in [contents, filename]:
+            raise PreventUpdate
+        if len(contents) > 1:
+            parse_multiple_contents(contents, filename)
+        else:
+            df = upload_string_to_df(*contents)
+            df[df < 0] = 0
+            df[~df.isin([np.nan, np.inf, -np.inf]).any(1)]
+            df = df.fillna(0)
+            return [
+                {"visibility": "visible"},
+                df.to_dict("records"),
+                [{"name": i, "id": i, "deletable": True, "selectable": True} for i in df.columns],
+            ]
+
+    @app.callback(
+        Output("data-table-visualization", "style_data_conditional"),
+        [Input("data-table-visualization", "selected_columns"), Input("data-table-visualization", "selected_rows"),],
+    )
+    def update_styles(selected_columns, selected_rows):
+        # Generates a list of dictionaries that style the selected rows and columns with a light blue highlight
+        return [
+            *[{"if": {"column_id": i}, "background_color": "#D2F3FF"} for i in selected_columns],
+            *[{"if": {"row_index": i}, "background_color": "#D2F3FF"} for i in selected_rows],
+        ]
+
+    @app.callback(
+        Output("upload-confirmation", "style"),
+        [Input("redis-upload-button", "n_clicks")],
+        [
+            State("data-table-visualization", "data"),
+            State("data-table-visualization", "selected_columns"),
+            State("data-table-visualization", "selected_rows"),
+        ],
+    )
+    def push_to_redis(button_clicks, data_dict, selected_columns, selected_rows):
+        if None in [button_clicks, data_dict, selected_columns]:
+            raise PreventUpdate
+
+        df = (
+            pd.DataFrame.from_records(data_dict)
+            if len(selected_rows) == 0
```

Comment:
I'd replace with `if not selected_rows:`

Reply:
addressed in 5a1d5c3